File: c_plus/Valid_Parentheses.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (auto ch : s) {
            if (ch == '{' || ch == '[' || ch == '(') {
                st.push(ch);
            }else if (ch == '}' || ch == ']' || ch == ')') {
                if (st.empty()) return false;
                char top = st.top();
                if ((top == '{' && ch == '}') || (top == '[' && ch == ']') || (top == '(' && ch == ')')){
                    st.pop();
                }else {
                    return false;
                }
            }else {
                return false;
            }
        }
        return st.empty();
    }
};

int main(){
    Solution s;
    cout << s.isValid("()") << endl;
}

File: py/Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        openList = ['[', '{', '(']
        closeList = [']', '}', ')']
        for i in s:
            if i in openList:
                stack.append(i)
            elif i in closeList:
                pos = closeList.index(i)
                if len(stack) > 0 and \
                    openList[pos] == stack[len(stack) - 1]:
                    stack.pop()
                else:
                    return False
        if len(stack) == 0:
            return True

if __name__ == '__main__':
    s = Solution()
    print(s.isValid(''))